File: ppggenerator.m
close all

noise=0.1;
DC=20;
AC=3;
DC2=10;
AC2=1;
DN=1;
fluctuation=1;
ff=1; %fluctuation frequency
samplingrate=120;
heartrate=60;
time=300;

t=0:1/samplingrate:time;
a=4*sin(t*2*pi*heartrate/60)+4;
b=sin(t*2*pi*heartrate/60+pi*1.4)*4-3.5;


for i=1:length(a)
    if a(i)<0
        a(i)=0;
    end
end

for i=1:length(b)
    if b(i)<0
        b(i)=0;
    end
end
    
c=AC*(a+DN*b*4)/10;
%---------Add noise

c=c+noise*rand(1,length(c))/2;

%---------Add DC

c=c+DC;
%--------Add fluctuation

if fluctuation ~= 0
    F=0.05*sin(ff*t/2);%+1;
    %c=fluctuation*c.*F;
    c1=fluctuation*c+10*F;
end
subplot(2,1,1)
plot(t,c1)
axis([0 max(t) 0 max(c1)])


%-------wave 2



t=0:1/samplingrate:time;
a=4*sin(t*2*pi*heartrate/60)+4;
b=sin(t*2*pi*heartrate/60+pi*1.4)*4-3.5;


for i=1:length(a)
    if a(i)<0
        a(i)=0;
    end
end

for i=1:length(b)
    if b(i)<0
        b(i)=0;
    end
end
    
c=AC2*(a+DN*b*4)/10;
%---------Add noise

c=c+noise*rand(1,length(c))/2;

%---------Add DC

c=c+DC2;
%--------Add fluctuation

if fluctuation ~= 0
    F=0.05*sin(ff*t/2);%+1;
    %c=fluctuation*c.*F;
    c2=fluctuation*c+10*F;
end

subplot(2,1,2)
plot(t,c2)
axis([0 max(t) 0 max(c2)])

File: PPGgenerator.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>

main()
{
    int samplingrate=60;
    int time=10; //Seconds
    int heartrate=60;
    int NrOfElements=samplingrate*time;
    int noise=1;
    int DC=15;
    int AC=1;
    int DN=1; //Dicrotic notch
    int fluctuation=1;
	float wave1[NrOfElements];
	float wave2[NrOfElements];
	float wave3[NrOfElements];
	float wave4[NrOfElements];
	int i;
	float pi=3.14159;
	int random;




	for ( i = 0; i <= NrOfElements; i++ )
    {
        wave1[i]=4*sin(i*2*pi*heartrate/(60*samplingrate))+4;
        if  ( wave1[i]<0 )
        {
            wave1[i]=0;
        }

        wave2[i]=4*sin(i*2*pi*heartrate/(60*samplingrate)+pi*1.4)-3.5;
         if  ( wave2[i]<0 )
        {
            wave2[i]=0;
        }
        float random=(rand() % 10)/10;
        wave3[i]=AC*(wave1[i]+DN*4*wave2[i])+noise*random/20+DC;
    }

//for ( i = 0; i <= 10; i++ )
//{
//	printf("%3.4f\n", wave3[i]);
//}
    if (fluctuation != 0)
    {
        for ( i = 0; i <= NrOfElements; i++ )
        {
            wave4[i]=0.05*sin(i/samplingrate)+1;
            wave3[i]=fluctuation*wave4[i]*wave3[i];
        }
    }


//for ( i = 0; i < 10; i++ )
//{
//	printf("%3.4f\n", wave3[i]);
//}

   return(0);
}
